routine/src/clj/parts/routine/core.clj

```diff
@@ -0,0 +1,136 @@
+(ns parts.routine.core
+  (:require [parts.routine.schedulers :as schedulers]))
+
+(defn add-routines
+  [w]
+  (assoc w
+         :routines
+         (filter
+           (fn [routine]
+             (when (:routine/kind routine)
+               (assert (:routine/fn routine))
+               (assert (:routine/interval-ms routine))
+               (assert (:routine/category routine))))
+           ((:system/get-register w)))))
+
+;; Example: run a task every 5 seconds with an initial 1 second delay
+(defn schedule-routine!
+  [w]
+  (let [routine (:routine w)]
+    (.scheduleWithFixedDelay (:scheduler w)
+                             (fn []
+                               (try
+                                 (:routine/fn routine)
+                                 (catch Throwable e
+                                   ;; If `scheduleWithFixedDelay`
+                                   ;; encounters an exception it
+                                   ;; suppresses subsequent
+                                   ;; executions. Therefore any errors
+                                   ;; are catched and logged here to
+                                   ;; keep executing the `routine-fn`,
+                                   ;; even if it contains a bug and does
+                                   ;; not catch all exceptions:
+                                   ((:log/error w
+                                      identity)
+                                    {:log/error :routine/unhandled-exception
+                                     :message "unhandled exception - a routine-fn should handle all exceptions"
+                                     :routine routine
+                                     :exception e})
+                                   )))
+                             0
+                             (:routine/interval-ms routine)
+                             java.util.concurrent.TimeUnit/MILLISECONDS)))
+
+(def termination-timeout (* 1000 30))
+
+(defn shutdown-and-await-termination
+  "Shutdowns a `java.util.concurrent.ExecutorService` gracefully. Based
+   on example `shutdownAndAwaitTermination` from
+   https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html"
+  [w]
+  ;; Disable new tasks from being submitted:
+  (let [^java.util.concurrent.ExecutorService scheduler (:scheduler w)]
+    (.shutdown scheduler)
+    (try
+      ;; Wait a while for existing tasks to terminate:
+      (when-not (.awaitTermination scheduler
+                                   termination-timeout
```

Comment:
Should be configurable like:

(or (::termination-timeout w)
      (* 1000 30))